syntax = "proto3";
package arike_pb;

import "arike_utils.proto";

message VariableEvent {
    Event event = 1;
    optional string str_value = 2;
    optional string str_low_limit = 3;
    optional string str_high_limit = 4;
    optional int64 int_value = 5;
    optional int64 int_low_limit = 6;
    optional int64 int_high_limit = 7;
    optional double float_value = 8;
    optional double float_low_limit = 9;
    optional double float_high_limit = 10;
    optional bool bool_value = 11;
    optional bool bool_low_limit = 12;
    optional bool bool_high_limit = 13;
}

message TsVariableMeta {
    string name = 1;
    ValType val_type = 2;
}

message TsVarValue {
    string name = 1;
    optional ValType val_type = 2;
    optional int64 int_value = 3;
    optional double float_value = 4;
    optional string str_value = 5;
    optional bool bool_value = 6;
    optional uint64 timestamp = 7;
}

message CreateVariablesRequest {
    string collection = 1;
    repeated TsVariableMeta variables = 2;
}

message CreateVariablesResponse {
    StatusCode status = 1;
    repeated string already_exists = 2;
}

message DeleteVariablesRequest {
    string collection = 1;
    repeated string names = 2;
}

message DeleteVariablesResponse {
    StatusCode status = 1;
    repeated string not_found = 2;
}

message ListVariablesRequest {
    string collection = 1;
    optional string pattern = 2;
}

message ListVariablesResponse {
    StatusCode status = 1;
    repeated TsVariableMeta variables = 2;
}

message SetVariablesRequest {
    string collection = 1;
    optional uint64 timestamp = 2;
    repeated TsVarValue values = 3;
}

message SetVariablesResponse {
    StatusCode status = 1;
    repeated string not_found = 2;
    repeated string invalid_type = 3;
}

message GetVariablesRequest {
    string collection = 1;
    repeated string names = 2;
}

message GetVariablesResponse {
    StatusCode status = 1;
    repeated TsVarValue values = 2;
}

message SubscribeVariablesRequest {
    string collection = 1;
    repeated string names = 2;
    repeated VariableEvent events = 3;
}
